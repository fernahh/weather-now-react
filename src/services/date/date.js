import getTime from 'date-fns/get_time'
import differenceInMilliseconds from 'date-fns/difference_in_milliseconds'
import format from 'date-fns/format'

const difference = (initialDate, lastDate) => {
  return differenceInMilliseconds(initialDate, lastDate)
}

const formatTimestamp = ({ timestamp, formatType}) => {
  return format(new Date(timestamp), formatType)
}

const now = () => getTime(new Date())

export default { difference, formatTimestamp, now }