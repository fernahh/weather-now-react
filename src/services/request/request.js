import axios from 'axios'

const get = (url, params) => params ? axios.get(url, { params }) : axios.get(url)

export default {
  get
}