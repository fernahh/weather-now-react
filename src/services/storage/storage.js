import date from '../date/date'

const get = (key) => localStorage.getItem(key)

const remove = (key) => localStorage.removeItem(key)

const set = (key, data) => localStorage.setItem(key, addUpdatedAt(data))

const addUpdatedAt = (data) => Object.assign(data, { updatedAt: date.now() })

export default {
  get,
  remove,
  set
}