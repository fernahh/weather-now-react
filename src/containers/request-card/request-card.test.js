import React from 'react'
import { shallow } from 'enzyme'
import addMilliseconds from 'date-fns/add_milliseconds'
import subMilliseconds from 'date-fns/sub_milliseconds'
import getTime from 'date-fns/get_time'
import RequestCard from './request-card'
import Card from '../../components/card/card'
import Loader from '../../components/loader/loader'
import Alert from '../../components/alert/alert'
import storage from '../../services/storage/storage'
import request from '../../services/request/request'
import date from '../../services/date/date'

jest.mock('../../services/request/request')
jest.mock('../../services/storage/storage')
jest.useFakeTimers()
date.now = jest.fn()

describe('<RequestCard />', () => {
  const onGetDataSuccess = jest.fn()
  const onGetDataError = jest.fn()

  const getRequestCardWrapper = (refreshInterval, cacheKey, cacheTime) => {
    return shallow(
      <RequestCard 
        cacheKey={cacheKey}
        cacheTime={cacheTime}
        title="Foobar" 
        url="http://api.com/dog"
        onGetDataSuccess={onGetDataSuccess}
        onGetDataError={onGetDataError}
        refreshInterval={refreshInterval}>
        <p className="content">Content</p>
      </RequestCard>
    )
  }

  describe('when render', () => {
    request.get.mockImplementation(() => new Promise(jest.fn())) 
    const wrapper = getRequestCardWrapper()

    it('render correctly', () => {
      expect(wrapper).toMatchSnapshot()
    })
    
    it('do not call request get method with url when data is cached', () => {
      const cacheKey = 'dog'
      const cacheTime = 6000
      const beforeCacheTime = 8000
      const beforeDate = getTime(subMilliseconds(getTime(new Date()), beforeCacheTime))
      const nowDate = getTime(subMilliseconds(beforeDate, beforeCacheTime))
      
      request.get.mockClear()
      storage.get.mockReturnValue({ name: 'Lessie', updateAt: beforeDate})
      date.now.mockReturnValue(nowDate)
      getRequestCardWrapper(null, cacheKey, cacheTime)

      expect(request.get).not.toBeCalled()
    })

    it('call request get method with url when cache is less than TTL', () => {
      const cacheKey = 'dog'
      const cacheTime = 120000
      const afterCacheTime = 130000
      const beforeDate = getTime(new Date())
      const nowDate = getTime(addMilliseconds(beforeDate, afterCacheTime))
      
      request.get.mockClear()
      storage.get.mockReturnValue({ name: 'Lessie', updateAt: beforeDate})
      date.now.mockReturnValue(nowDate)
      getRequestCardWrapper(null, cacheKey, cacheTime)
      
      expect(request.get).toHaveBeenCalledWith('http://api.com/dog', {})
    })
    
    it('call request get method with url', () => {
      storage.get.mockReturnValue(false)
      getRequestCardWrapper(6000, 'dog', 7000)
      expect(request.get).toHaveBeenCalledWith('http://api.com/dog', {})
    })
  })

  describe('when get data complete with success', () => {
    const data = { name: 'Lassie' }
    request.get.mockImplementation(() => Promise.resolve(data))
    const wrapper = getRequestCardWrapper()

    it('render correctly', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('execute onGetDataSuccess callback with data response', () => {
      expect(onGetDataSuccess).toHaveBeenCalledWith(data)
    })
  })

  describe('when get data is rejected', () => {
    const error = { message: 'Some error message' }
    request.get.mockImplementation(() => Promise.reject(error))
    const wrapper = getRequestCardWrapper()

    it('render correctly', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('execute onGetDataError callback with error response', () => {
      expect(onGetDataError).toHaveBeenCalledWith(error)
    })
  })

  describe('when refresh interval', () => {
    request.get.mockImplementation(() => new Promise(jest.fn())) 
    const wrapper = getRequestCardWrapper(6000)

    it('call get method with url', () => {
      request.get.mockClear()
      expect(request.get).not.toBeCalled()
    
      jest.runOnlyPendingTimers()
      expect(request.get).toHaveBeenCalledWith('http://api.com/dog', {})
    })

    it('clear interval on unmount', () => {
      wrapper.unmount()
      expect(clearInterval).toHaveBeenCalled()
    })
  })

  describe('when refresh interval is not declared', () => {
    request.get.mockImplementation(() => new Promise(jest.fn())) 
    const wrapper = getRequestCardWrapper()

    it('do not call set interval method', () => {
      setInterval.mockClear()
      expect(setInterval).not.toBeCalled()
    })

    it('do not call clear interval on unmount', () => {
      clearInterval.mockClear()
      wrapper.unmount()
      expect(clearInterval).not.toBeCalled()
    })
  })

  describe('when cache response data', () => {
    const data = { name: 'Lassie' }
    getRequestCardWrapper(6000, 'dog', 7000)
    request.get.mockImplementation(() => Promise.resolve(data))

    it('set item on storage', () => {
      expect(storage.set).toHaveBeenCalledWith('dog', data)
    })
  })
})