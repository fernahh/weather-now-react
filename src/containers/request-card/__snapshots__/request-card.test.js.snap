// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestCard /> when get data complete with success render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RequestCard
    cacheKey={undefined}
    cacheTime={undefined}
    errorMessage="Something went wrong"
    onGetDataError={[Function]}
    onGetDataSuccess={[Function]}
    refreshInterval={undefined}
    title="Foobar"
    url="http://api.com/dog"
>
    <p
        className="content"
    >
        Content
    </p>
</RequestCard>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <Card
        title="Foobar"
>
        <Loader />
</Card>,
      "className": "request-card",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Loader />,
          false,
          false,
        ],
        "title": "Foobar",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
        false,
      ],
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Card
          title="Foobar"
>
          <Loader />
</Card>,
        "className": "request-card",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <Loader />,
            false,
            false,
          ],
          "title": "Foobar",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          false,
          false,
        ],
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`<RequestCard /> when get data is rejected render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RequestCard
    cacheKey={undefined}
    cacheTime={undefined}
    errorMessage="Something went wrong"
    onGetDataError={[Function]}
    onGetDataSuccess={[Function]}
    refreshInterval={undefined}
    title="Foobar"
    url="http://api.com/dog"
>
    <p
        className="content"
    >
        Content
    </p>
</RequestCard>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <Card
        title="Foobar"
>
        <Loader />
</Card>,
      "className": "request-card",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Loader />,
          false,
          false,
        ],
        "title": "Foobar",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
        false,
      ],
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Card
          title="Foobar"
>
          <Loader />
</Card>,
        "className": "request-card",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <Loader />,
            false,
            false,
          ],
          "title": "Foobar",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          false,
          false,
        ],
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`<RequestCard /> when render render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RequestCard
    cacheKey={undefined}
    cacheTime={undefined}
    errorMessage="Something went wrong"
    onGetDataError={[Function]}
    onGetDataSuccess={[Function]}
    refreshInterval={undefined}
    title="Foobar"
    url="http://api.com/dog"
>
    <p
        className="content"
    >
        Content
    </p>
</RequestCard>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <Card
        title="Foobar"
>
        <Loader />
</Card>,
      "className": "request-card",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Loader />,
          false,
          false,
        ],
        "title": "Foobar",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
        false,
      ],
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Card
          title="Foobar"
>
          <Loader />
</Card>,
        "className": "request-card",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <Loader />,
            false,
            false,
          ],
          "title": "Foobar",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          false,
          false,
        ],
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
